### ISSUE TITLE
训练时内存占用一直增加

### ISSUE THREAD

**Issue by jim66666** [VISION-SJTU/DiffusionTrack#2]
作者大大，我在单gpu服务器上训练模型的时候，内存占用一直增加，直到最后进程会被kill掉，请问这是怎么回事？我应该如何修改代码呢？

**Comment by zxyyxzz** [comment]
兄弟，解决了吗？

**Comment by zxyyxzz** [comment]
> 作者大大，我在单gpu服务器上训练模型的时候，内存占用一直增加，直到最后进程会被kill掉，请问这是怎么回事？我应该如何修改代码呢？

兄弟解决了吗？是什么问题？

**Comment by jim66666** [comment]
没有，还需要排查，你也遇到这个问题了吗

**Comment by zxyyxzz** [comment]
> 没有，还需要排查，你也遇到这个问题了吗

遇到了，作者代码应该有问题，应该是哪个list里一直在往里加东西

**Comment by jim66666** [comment]
慢慢找吧，或者等作者更新代码

**Comment by zxyyxzz** [comment]
> 慢慢找吧，或者等作者更新代码

我今天必须找到，急用

**Comment by jim66666** [comment]
实在不行就只能加载checkpoint训练了，祝你好运

**Comment by zxyyxzz** [comment]
> 实在不行就只能加载checkpoint训练了，祝你好运

什么意思？加载checkpoint训练？

**Comment by jim66666** [comment]
如果不能解决内存占用问题还想要训练模型，就在每个epoch结束后都保存checkpoint，如果训练终止了，就加载上一个checkpoint接着训练。不过这样做效率很低，要按照作者设计的训练200epoch估计要很久，我也没有什么其他方法了。如果你能解决内存占用问题，请告诉我一声，感谢。

**Comment by zxyyxzz** [comment]
> 如果不能解决内存占用问题还想要训练模型，就在每个epoch结束后都保存checkpoint，如果训练终止了，就加载上一个checkpoint接着训练。不过这样做效率很低，要按照作者设计的训练200epoch估计要很久，我也没有什么其他方法了。如果你能解决内存占用问题，请告诉我一声，感谢。

好，我找找

**Comment by zxyyxzz** [comment]
> 如果不能解决内存占用问题还想要训练模型，就在每个epoch结束后都保存checkpoint，如果训练终止了，就加载上一个checkpoint接着训练。不过这样做效率很低，要按照作者设计的训练200epoch估计要很久，我也没有什么其他方法了。如果你能解决内存占用问题，请告诉我一声，感谢。

我好像找到问题了，我先训一下，看看还会不会爆显存，解决了的话就告诉你。

**Comment by zxyyxzz** [comment]
> 如果不能解决内存占用问题还想要训练模型，就在每个epoch结束后都保存checkpoint，如果训练终止了，就加载上一个checkpoint接着训练。不过这样做效率很低，要按照作者设计的训练200epoch估计要很久，我也没有什么其他方法了。如果你能解决内存占用问题，请告诉我一声，感谢。

我好像找到问题了，我先训一下，看看还会不会爆内存，解决了的话就告诉你。

**Comment by jim66666** [comment]
好的

**Comment by zxyyxzz** [comment]
> 好的

找到了，是他记录loss的时候没有用item()，tensor被不断存储，将lib\train\actors\diffusiontrack.py文件中的compute_losses函数改一下就行了：
`    def compute_losses(self, outputs, new_targets, return_status=True):
        loss_dict, weight_dict = self.net(outputs = outputs, new_targets = new_targets, mode = 'compute_loss')
        new_loss_dict= {}
        for k in loss_dict.keys():
            if k in weight_dict:
                new_loss_dict[k] = loss_dict[k] * weight_dict[k]

        losses = sum(new_loss_dict.values())
        status = loss_dict.copy()
       ###我加了下面两行####
       ###################
        for k, v in status.items():
            status[k] = v.item()
        ###################
        status['total_loss'] = losses.clone().detach().item()
        if return_status:
            # status for log
            return losses, status
        else:
            return losses`

**Comment by jim66666** [comment]
感谢你的帮助，大善人

**Comment by zxyyxzz** [comment]
> 感谢你的帮助，大善人
有用不，我刚刚又不太确定了。。。。

**Comment by jim66666** [comment]
我现在没有服务器，暂时无法验证

**Comment by zxyyxzz** [comment]
> 我现在没有服务器，暂时无法验证

我现在确定有用，work了。

**Comment by TengXiang164** [comment]
大佬们，是哪个list出现的问题呀？
